Add a declarative version of the protect-form component

jsnajdr opened the pull request:
Added a `ProtectForm` component that can be used instead of the HOC version in a more declarative fashion, suitable for integration with Redux state.

The HOC version, with its imperative `markChanged` and `markSaved` methods, is one of obstacles when converting components to use Redux and the data layer.

I used the new component immediately to convert the `AccountPassword` form, to show an example usage.

spen commented on client/lib/protect-form/index.jsx:
```diff
@@ -57,6 +67,41 @@ export const protectForm = WrappedComponent => {
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'beforeunload', this.warnIfChanged );
```
> Is this safe in terms of memory

Ahh, of course ;) FFTI my previous comment!

dmsnell replied:
what is FFTI?

spen replied:
Sorry, 'Feel Free To Ignore'.
I'd noticed that we were removing this listener after making the previous comment on line 80.
In my first pass I hadn't connected the dots... I could have been a bit more clear that it was that comment that could be ignored :)

spen commented on client/lib/protect-form/README.md:
````diff
@@ -1,34 +1,73 @@
+... and then render the `ProtectForm` inside your form to protect it, while changing
+component state instead of calling functions:
+
+```js
````
Could also be `jsx`

dmsnell commented on client/lib/protect-form/README.md:
````diff
@@ -1,34 +1,73 @@
+The `protectForm` HoC provides a `markChanged` prop you can call on the `onChange` event of your form like this:
 
-```jsx
+```js
````
we don't want/need to be using `jsx` on these code fences

spen replied:
We would benefit from more specific syntax highlighting, but ` ```js` might be more future proof.

jsnajdr replied:
GitHub does a better syntax highlighting for JSX when ` ```jsx` is used instead of ` ```js`, so I added it at the two places where it matters.

dmsnell commented on client/lib/protect-form/index.jsx:
```diff
@@ -15,8 +15,37 @@ const confirmText = i18n.translate( 'You have unsaved changes. Are you sure you
 	return class ProtectedFormComponent extends Component {
+		// This function binding makes the listener attached to window unique for each
+		// component instance.
+		warnIfChanged = warnIfChanged.bind();
```
we should probably never call `bind` without any parameters. if we want no binding then we should supply `null` like below. if we want another specific binding, we should pass it in.

am I missing something here?

jsnajdr replied:
The only purpose of this `bind` call was to create an unique function object, so that `addEventListener`/`removeEventListener` work correctly. But I rewrote this code completely, there's no longer a need to call `bind` here.

dmsnell commented on client/lib/protect-form/index.jsx:
```diff
@@ -15,8 +15,37 @@ const confirmText = i18n.translate( 'You have unsaved changes. Are you sure you
+		warnIfChanged = warnIfChanged.bind();
+
+		markChanged = markChanged.bind( null, this );
+		markSaved = markSaved.bind( null, this );
```
this is an odd construction the should probably have an explanation. usually one would expect `this` to be in the first argument but here it's being explicitly set so that the `this` in `markChanged` and `markSaved` is `null`/global but the argument `form` is equal to the `this` here at the bind-site.

maybe this is evidence that we could rewrite these functions to simplify things a bit. did we consider using arrow-functions to preserve lexical binding of `this`?

jsnajdr replied:
You're right, I rewrote the `bind` calls to arrow functions here. Much cleaner now.

dmsnell commented on client/lib/protect-form/index.jsx:
```diff
@@ -57,6 +67,41 @@ export const protectForm = WrappedComponent => {
+
+	// This function binding makes the listener attached to window unique for each
+	// component instance.
+	warnIfChanged = warnIfChanged.bind();
```
@spen there are no memory concerns as long as we don't end up with loops that create these components. we mainly want to avoid `bind` inside of React props because those get recreated on every render. also it's not limited to `bind` but arrow functions and closures of any kind.

jsnajdr replied:
I rewrote the code that adds and removes the `beforeunload` listener to add the listener only once, and reference-count multiple instances. It should be much cleaner now, and there is only one `warnIfChanged` function.

dmsnell commented on client/lib/protect-form/index.jsx:
```diff
@@ -57,6 +67,41 @@ export const protectForm = WrappedComponent => {
+	render() {
+		return null;
+	}
+}
```
does this component not render any wrapped components?

jsnajdr replied:
It doesn't have any children -- it's just a guard component that does stuff on mount/unmount/prop change, and renders nothing. Very similar to, e.g., `<QueryGeo />`.

jsnajdr commented on client/lib/protect-form/index.jsx:
```diff
@@ -57,6 +67,41 @@ export const protectForm = WrappedComponent => {
+/*
+ * Declarative variant that takes a 'isChanged' prop.
+ */
+export class ProtectForm extends Component {
```
> Longer term, do we plan to keep both?

I don't know at this moment. The declarative component works much better with Redux, but the HOC one may remain useful for some cases forever.

Anyway, both components will have to live next to each other for some time, so I renamed the declarative one to `ProtectFormGuard` to avoid naming confusion.

jsnajdr commented:
Pushed a new version with the following changes:
- renamed the component from `ProtectForm` to `ProtectFormGuard`
- rewrote the `beforeunload` listener code to attach the listener only once and refcount instances
- rewrote confusing calls to `bind` to arrow functions
- added ` ```jsx` to docs at places where it leads to better syntax highlighting

gwwar commented:
Was there anything blocking here? Otherwise we can probably retest and 🚢

jsnajdr commented:
@gwwar I rebased and retested PR and everything is ready to :ship:. If you or @dmsnell give me a ✅ I'll merge.

gwwar commented on client/lib/protect-form/index.jsx:
```diff
@@ -14,37 +14,58 @@ const debug = debugModule( 'calypso:protect-form' );
+
+function addBeforeUnloadListener() {
+	if ( listenerCount === 0 ) {
+		window.addEventListener( 'beforeunload', warnIfChanged );
```
This is minor but I'd recommend checking for window before we call it. If we ever want to test this, it can be a pain to work around.

jsnajdr commented:
Added existence checks for `window` and going to merge.
